File: term7/TK/Fano.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Node
{
    public char Symbol { get; set; }
    public double Probability { get; set; }
    public Node Left { get; set; }
    public Node Right { get; set; }
}

class Program
{
    static Node BuildFanoTree(List<Node> symbols)
    {
        if (symbols.Count == 1)
            return symbols[0];

        symbols = symbols.OrderByDescending(s => s.Probability).ToList();

        return BuildFanoTreeHelper(symbols);
    }

    static Node BuildFanoTreeHelper(List<Node> symbols)
    {
        if (symbols.Count == 1)
            return symbols[0];

        double totalProbability = symbols.Sum(s => s.Probability);
        double targetProbability = totalProbability / 2;
        double currentProbability = 0;
        int splitIndex = 0;

        for (int i = 0; i < symbols.Count; i++)
        {
            currentProbability += symbols[i].Probability;
            if (currentProbability >= targetProbability)
            {
                splitIndex = i;
                break;
            }
        }

        List<Node> leftSymbols = symbols.GetRange(0, splitIndex + 1);
        List<Node> rightSymbols = symbols.GetRange(splitIndex + 1, symbols.Count - splitIndex - 1);

        Node leftSubtree = BuildFanoTreeHelper(leftSymbols);
        Node rightSubtree = BuildFanoTreeHelper(rightSymbols);

        Node root = new Node { Symbol = '\0', Probability = totalProbability };
        root.Left = leftSubtree;
        root.Right = rightSubtree;

        return root;
    }

    static Dictionary<char, string> EncodeFanoTree(Node root, string code = "")
    {
        if (root.Symbol != '\0')
            return new Dictionary<char, string> { { root.Symbol, code } };

        Dictionary<char, string> codeMap = new Dictionary<char, string>();
        codeMap = codeMap.Concat(EncodeFanoTree(root.Left, code + "0"))
                         .Concat(EncodeFanoTree(root.Right, code + "1"))
                         .ToDictionary(x => x.Key, x => x.Value);

        return codeMap;
    }

    static void Main()
    {
        string text = "Hello, world!";
        Dictionary<char, double> symbolCounts = new Dictionary<char, double>();

        foreach (char symbol in text)
        {
            if (symbolCounts.ContainsKey(symbol))
                symbolCounts[symbol]++;
            else
                symbolCounts[symbol] = 1;
        }

        List<Node> symbols = symbolCounts.Select(kv => new Node { Symbol = kv.Key, Probability = kv.Value / text.Length }).ToList();
        Node fanoTree = BuildFanoTree(symbols);
        Dictionary<char, string> codeMap = EncodeFanoTree(fanoTree);

        foreach (var kvp in codeMap)
        {
            Console.WriteLine($"Symbol: {kvp.Key}, Code: {kvp.Value}");
        }
    }
}

File: term7/TK/Fano.py
from heapq import heappop, heappush

class Node:
    def __init__(self, symbol, probability):
        self.symbol = symbol
        self.probability = probability
        self.left = None
        self.right = None

def build_fano_tree(symbols):
    if len(symbols) == 1:
        return symbols[0]

    # Отсортировать символы по убыванию вероятностей
    symbols = sorted(symbols, key=lambda x: x.probability, reverse=True)

    # Рекурсивно построить дерево Фано
    return build_fano_tree_helper(symbols)

def build_fano_tree_helper(symbols):
    if len(symbols) == 1:
        return symbols[0]

    # Найти точку разделения алфавита
    total_probability = sum(symbol.probability for symbol in symbols)
    target_probability = total_probability / 2
    current_probability = 0
    split_index = None

    for i in range(len(symbols)):
        current_probability += symbols[i].probability
        if current_probability >= target_probability:
            split_index = i
            break

    # Разделить алфавит на две части
    left_symbols = symbols[:split_index + 1]
    right_symbols = symbols[split_index + 1:]

    # Рекурсивно построить левое и правое поддерево
    left_subtree = build_fano_tree_helper(left_symbols)
    right_subtree = build_fano_tree_helper(right_symbols)

    # Создать узел, соединяющий левое и правое поддерево
    root = Node(None, total_probability)
    root.left = left_subtree
    root.right = right_subtree

    return root

def encode_fano_tree(root, code=''):
    if root.symbol is not None:
        # Листовой узел
        return {root.symbol: code}
    else:
        # Внутренний узел
        code_map = {}
        code_map.update(encode_fano_tree(root.left, code + '0'))
        code_map.update(encode_fano_tree(root.right, code + '1'))
        return code_map

# Пример использования
text = "Hello, world!"
symbol_counts = {}

for symbol in text:
    symbol_counts[symbol] = symbol_counts.get(symbol, 0) + 1

symbols = [Node(symbol, count / len(text)) for symbol, count in symbol_counts.items()]
fano_tree = build_fano_tree(symbols)
code_map = encode_fano_tree(fano_tree)

for symbol, code in code_map.items():
    print(f"Symbol: {symbol}, Code: {code}")